### PR TITLE
[rllib] Add PPO+DQN two trainer multiagent workflow example

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Adds an example of using the new training workflow api to compose concurrent multi-agent training. In this example, PPO is trained with a certain batch size and DQN is trained with a different one. They both share the same stream of experiences.

This is purposely not added to the rst documentation yet.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/examples/two_trainer_workflow.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+from ray.rllib.optimizers.async_replay_optimizer import LocalReplayBuffer
+from ray.tune.registry import register_env
+from ray.util.iter import LocalIterator
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--num-iters", type=int, default=20)
+
+
+def custom_training_workflow(workers, config):
+    local_replay_buffer = LocalReplayBuffer(
```
Comment (ericl) [inline rllib/examples/two_trainer_workflow.py:34]: @richardliaw any specific comments on confusing bits I'm this function, and why?

File: rllib/agents/dqn/apex.py [hunk 2]
```diff
@@ -84,6 +84,32 @@ def update_target_based_on_num_steps_trained(trainer, fetches):
...
+# Update worker weights as they finish generating experiences.
+class UpdateWorkerWeights:
+    def __init__(self, learner_thread, workers, max_weight_sync_delay):
+        self.learner_thread = learner_thread
+        self.workers = workers
+        self.steps_since_update = collections.defaultdict(int)
+        self.max_weight_sync_delay = max_weight_sync_delay
+        self.weights = None
+
+    def __call__(self, item):
```
Comment (richardliaw) [inline rllib/agents/dqn/apex.py:96]: nit/suggestion: What if you type hint these args? One question I have when reviewing is what are these args, and how do I know where to look if I wanted to add arguments/parameters?

File: rllib/examples/two_trainer_workflow.py (hunk 1 above)
Comment (richardliaw) [reply rllib/examples/two_trainer_workflow.py:34]: yeah - as a user, the first question I have is, what is "args - workers"? is there a docstring or type hint you can add here so I can reference in the docs

File: rllib/examples/two_trainer_workflow.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        num_shards=1,
+        learning_starts=1000,
+        buffer_size=50000,
```
Comment (richardliaw) [inline rllib/examples/two_trainer_workflow.py:37]: maybe comment here what the units are (learning_starts= 1000 steps? batches? samples?)

Comment (richardliaw) [reply rllib/examples/two_trainer_workflow.py:37]: same with buffer size

File: rllib/examples/two_trainer_workflow.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        replay_batch_size=64)
+
+    def add_ppo_metrics(batch):
```
Comment (richardliaw) [inline rllib/examples/two_trainer_workflow.py:40]: type hint would be helpful here.

File: rllib/examples/two_trainer_workflow.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        metrics = LocalIterator.get_metrics()
+        metrics.counters["agent_steps_trained_PPO"] += batch.total()
+        return batch
+
+    def add_dqn_metrics(batch):
+        print("DQN policy learning on samples from",
+              batch.policy_batches.keys(), "env steps", batch.count,
+              "agent steps", batch.total())
+        metrics = LocalIterator.get_metrics()
+        metrics.counters["agent_steps_trained_DQN"] += batch.total()
```
Comment (richardliaw) [inline rllib/examples/two_trainer_workflow.py:53]: what does LocalIterator do?

Comment (richardliaw) [reply rllib/examples/two_trainer_workflow.py:53]: one thing I'm confused about is when does the state of the local iterator get updated

File: rllib/examples/two_trainer_workflow.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        return batch
+
+    # Generate common experiences.
+    rollouts = ParallelRollouts(workers, mode="bulk_sync")
+    r1, r2 = rollouts.duplicate(n=2)
```
Comment (richardliaw) [inline rllib/examples/two_trainer_workflow.py:58]: is rollouts a stream?

Comment (ericl) [reply rllib/examples/two_trainer_workflow.py:58]: Yes. The assumption here is the user knows what's going on at least (that everything is a stream).

File: rllib/examples/two_trainer_workflow.py (hunk 1 above)
Comment (ericl) [reply rllib/examples/two_trainer_workflow.py:35]: Ok I'll add some type hints as possible for each operator.

Comment (ericl) [reply rllib/examples/two_trainer_workflow.py:35]: Not all in this PR though, it's a big change.

File: rllib/examples/two_trainer_workflow.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    dqn_store_op = r1.for_each(SelectExperiences(["dqn_policy"])) \
+        .for_each(
+            StoreToReplayBuffer(local_buffer=local_replay_buffer))
+    dqn_replay_op = Replay(local_buffer=local_replay_buffer) \
+        .for_each(add_dqn_metrics) \
+        .for_each(TrainOneStep(workers, policies=["dqn_policy"])) \
+        .for_each(UpdateTargetNetwork(
+            workers, target_update_freq=500, policies=["dqn_policy"]))
+    dqn_train_op = Concurrently(
+        [dqn_store_op, dqn_replay_op], mode="round_robin", output_indexes=[1])
```
Comment (richardliaw) [inline rllib/examples/two_trainer_workflow.py:71]: haha..... this just feels like TF1.0...

**Comment by richardliaw** [comment]
Generally looks good. I think *functional* operators make a lot of sense, though *stateful* operators I still find a hard time grasping.

In the example, it'd be nice to have a "inference" part after `tune.run` -- "take the trained policy and run it on some env". From a first glance, I don't know how I would do that.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**